### ISSUE TITLE
Typst: No table caption without label chunk option

### ISSUE THREAD

**Issue by vincentarelbundock** [quarto-dev/quarto-cli#9369]
### Bug description

In a Typst document, Quarto does not display a table's caption unless the `label` chunk option is defined.

Minimal reproducible example using the `tinytable` package to generate Typst tables:

[example.pdf](https://github.com/quarto-dev/quarto-cli/files/14970823/example.pdf)

````
---
format:
  typst:
    keep-typ: true
---

Caption does *not* appear:

```{r}
#| tbl-cap: "Fist table without label"

library(tinytable)
x <- mtcars[1:2, 1:2]

tt(x)
```

Caption appears correctly:

```{r}
#| label: tbl-tiny
#| tbl-cap: "Second table with label"
tt(x)
```
````

I attach the resulting PDF and paste the Typst table blocks below:

#### Table bad

```typst
#block[
#import "@preview/tablex:0.0.8": tablex, hlinex, vlinex, colspanx
#let nhead = 1;
#let nrow = 2;
#let ncol = 2;

#figure(
  
  kind: "tinytable",
  supplement: none,
  tablex(
    columns: ncol,
    header-rows: nhead,
    align: left + horizon,
    auto-lines: false,

hlinex(y: 0, start: 0, end: 2, stroke: 0.1em + black),
hlinex(y: 1, start: 0, end: 2, stroke: 0.05em + black),
hlinex(y: 3, start: 0, end: 2, stroke: 0.1em + black),
    // tinytable lines before

    map-cells: cell => {

      // tinytable cell style before
      return cell;
    },

    // tinytable cell content after
[mpg], [cyl],
[21], [6],
[21], [6],

  ) // end tablex
) // end figure
]
]
```

#### Table good

```typst
#figure([
#block[
#import "@preview/tablex:0.0.8": tablex, hlinex, vlinex, colspanx
#let nhead = 1;
#let nrow = 2;
#let ncol = 2;

#figure(
  
  kind: "tinytable",
  supplement: none,
  tablex(
    columns: ncol,
    header-rows: nhead,
    align: left + horizon,
    auto-lines: false,

hlinex(y: 0, start: 0, end: 2, stroke: 0.1em + black),
hlinex(y: 1, start: 0, end: 2, stroke: 0.05em + black),
hlinex(y: 3, start: 0, end: 2, stroke: 0.1em + black),
    // tinytable lines before

    map-cells: cell => {

      // tinytable cell style before
      return cell;
    },

    // tinytable cell content after
[mpg], [cyl],
[21], [6],
[21], [6],

  ) // end tablex
) // end figure
]
], caption: figure.caption(
position: top, 
[
Second table with label
]), 
kind: "quarto-float-tbl", 
supplement: "Table", 
numbering: "1", 
)
<tbl-tiny>
```

**Comment by mcanouil** [comment]
It seems to be an issue with your package as you take over the table, caption is not added.
Look at the code your function produces in the intermediate markdown.

````qmd
---
title: "Quarto Playground"
format: typst
keep-md: true
---

This is a playground for Quarto.

```{r}
library(tinytable)
x <- mtcars[1:2, 1:2]
```

```{r}
#| tbl-cap: This is a table
knitr::kable(x)
```

```{r}
#| tbl-cap: This is a table
tt(x)#| 
```
````

**Comment by vincentarelbundock** [comment]
The tinytable package produces the *exact same* Typst code in both of the examples in my initial Quarto notebook. The package is not aware of Quarto chunk options at all, and does not interact with them.

What differs in the two examples in my Quarto document is *not* the tinytable-generated code, is the the code that Quarto wraps around it.

And yes, tinytable does take over table generation. That's desirable.

What is undesirable is that Quarto wraps different code around Typst tables when the label chunk option is or is not defined.

**Comment by mcanouil** [comment]
Yes, but it's not detected as float object by Quarto unless it goes through the cross-reference code pat which the `tbl-` label prefix triggers.

The caption is simply not caught from markdown to Typst in the case of raw blocks, you can see that with `keep-md: true`.

<table>
<tr><th>knitr</th><th>tinytable</th></tr>
<tr><td>

````qmd
---
title: "Quarto Playground"
format: typst
keep-md: true
---

```{r}
library(knitr)
x <- mtcars[1:2, 1:2]
```

```{r}
#| tbl-cap: This is a table
kable(x)
```

```{r}
#| label: tbl-foo
#| tbl-cap: This is a table
kable(x)
```

```{r}
#| label: foo
#| tbl-cap: This is a table
kable(x)
```
````

</td><td>


````qmd
---
title: "Quarto Playground"
format: typst
keep-md: true
---

```{r}
library(tinytable)
x <- mtcars[1:2, 1:2]
```

```{r}
#| tbl-cap: This is a table
tt(x)
```

```{r}
#| label: tbl-foo
#| tbl-cap: This is a table
tt(x)
```

```{r}
#| label: foo
#| tbl-cap: This is a table
tt(x)
```
````

</td></tr>
</table>

**Comment by vincentarelbundock** [comment]
Thanks for the investigation, I appreciate it.

I understand what you are saying: The `tbl-` label triggers the cross-reference code that will wrap a raw code block in an environment that includes both the label reference and the caption. Thus, captions are only displayed for raw code block if a `tbl-` label is defined.

What I am saying is that this behavior is extremely counter-intuitive.

When would any user use the Quarto-specific `#| tbl-cap` code chunk syntax, and *not* expect Quarto to handle and display a caption? I would argue: Never. If a user supplies both `#| tbl-cap: Blah blah` and some raw code block, they almost certainly expect Quarto to display the "Blah blah" caption above whatever is in the raw code block, regardless of whether the document includes cross-references or not.

In other words, I think that the cross-ref wrapping code should be triggered whenever the user specifies a label OR a caption.

**Comment by mcanouil** [comment]
> When would any user use the Quarto-specific #| tbl-cap code chunk syntax, and not expect Quarto to handle and display a caption? I would argue: Never. If a user supplies both #| tbl-cap: Blah blah and some raw code block, they almost certainly expect Quarto to display the "Blah blah" caption above whatever is in the raw code block, regardless of whether the document includes cross-references or not.

As I said, that is the issue because this only happens in Typst.

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/6307b6d5-8fa8-4ae2-956b-e66dc2231b36">

**Comment by vincentarelbundock** [comment]
Ok ok, I got it. You were just confirming that this is a Typst-specific inconsistency. 

Sorry for misreading.

**Comment by cscheid** [comment]
Let me first just make sure everyone is on the same page.

## Don't emit `#figure()` if you're targeting Quarto documents

> And yes, tinytable does take over table generation. That's desirable.

It's desirable that `tinytable` takes over _table_ generation. But it's **not** desirable that it takes over crossref generation. If you do so, then half of the Quarto features will never work with your package. You can choose to do anything you want, of course. But, just to make this clear, we will consider it a bug on your side and we will not change our code.

If you're going to emit raw typst code to be consumed inside a Quarto document, then you'll need to make sure that your typst code conforms to Quarto's expectations. In general, do _not_ attempt to emit the `#figure...()` environment that surrounds the `tablex` call, because Quarto needs to reconfigure the crossreferencing environment.

The reason for that is that you don't know whether that "figure" will be referenced as a Figure, a Table, a Supplemental Table, etc. Better to just let Quarto take care of that. You already see part of the problem there when you say `kind: "tinytable"`.

Consider this minimal .qmd file:

````
---
title: Hello
format: typst
keep-typ: true
keep-md: true
---


```{r}
#| label: tbl-hello
#| tbl-cap: A caption.
#| echo: false

cat("This is actually a table")
```

See @tbl-hello.
````

This produces the following .typ block for the code cell:

````
#figure([
#block[
```
This is actually a table
```

]
], caption: figure.caption(
position: top, 
[
A caption.
]), 
kind: "quarto-float-tbl", 
supplement: "Table", 
numbering: "1", 
)
<tbl-hello>
````

Note `kind: "quarto-float-tbl"` specifically. This signals to Typst's crossref system that the contents of the figure should be numbered under a `quarto-float-tbl` counter. If you emit `kind: "tinytable"` in your figure, they will never be referenced with ours. Your code is also unlikely to know enough about the document configuration to be able to decide whether the caption should be positioned on `top` or `bottom`, and that needs to be emitted into the `#figure()` block.

tl;dr: your package doesn't have enough information about Quarto to emit `#figure()` compatibly. Focus instead on emitting the contents and let Quarto do its job.

## What to do with captions without labels

The problem here is that captions of non-crossreferenceable floats do not have a uniform representation in a number of different formats. Quarto tries to be consistent across formats, but is limited by the crossref support in some of the target formats. The problem you're seeing is not unique to Typst, it happens in LaTeX as well. There, tables with captions (and no label) end up counting against the `table` counter. If we choose to do the same in Typst, then we'll have to emit a `#figure()` block as well, and know that whatever Typst content is inside, it belongs in the `kind: "quarto-float-tbl"`. Nevertheless, Quarto will need to again be the one taking responsibility for emitting this code, because your library doesn't know the rest of what's happening with the crossref system.

**Comment by cscheid** [comment]
> I'm still not sure I understand the issue with captions and no labels. My library does not try to handle either of those.

Captions in typst work by adding `caption: ...` to `#figure()`. `#figure()` requires a `kind` field, which controls what counter is used. Using that same example as above to illustrate, note the `caption: figure.caption(...)` attribute:

````
#figure([
#block[
```
This is actually a table
```

]
], caption: figure.caption(
position: top, 
[
A caption.
]), 
kind: "quarto-float-tbl", 
supplement: "Table", 
numbering: "1", 
)
<tbl-hello>
````

The problem here is that if you don't give us a crossref id ("tbl-", or "fig-", or "lst-", etc), then we cannot even in principle know which `kind` to use.

**Comment by cscheid** [comment]
Ok, after reading all of this again, I really don't think there's a bug here.

Consider the example @mcanouil shared:

````
::: {.cell tbl-cap="This is a table where the caption is not there"}

```{=typst}
#figure(
  align(center)[#table(
    columns: 3,
    align: (left,right,right,),
    table.header([], [mpg], [cyl],),
    table.hline(),
    [Mazda RX4], [21], [6],
    [Mazda RX4 Wag], [21], [6],
  )]
  , kind: table
  )
```

:::
````

The problem here is that the raw typst code is taking over _crossref_ generation. This is not a good idea and it's not something Quarto can support for the reasons I gave above. Instead, the following should be used instead:

````
---
format: typst
---

::: {#tbl-1}

```{=typst}
  #align(center)[#table(
    columns: 3,
    align: (left,right,right,),
    table.header([], [mpg], [cyl],),
    table.hline(),
    [Mazda RX4], [21], [6],
    [Mazda RX4 Wag], [21], [6],
  )]
```

This is a table where the caption is there.

:::
````

This currently triggers a separate bug, https://github.com/quarto-dev/quarto-cli/issues/9887, but the crossref generation happens correctly.

**Comment by vincentarelbundock** [comment]
Thanks a lot for reading this report so closely and for the detailed answer. I agree with your conclusion and thank you for your help. Sorry for the time waste!
